### PR TITLE
Redis integration

### PR DESCRIPTION
**Pull request by ajvondrak**
This implementation monkey-patches the `Redis::Client#process` method
that powers all the other `Redis::Client` methods who make requests to
the Redis server:

- `call`
- `call_loop`
- `call_pipeline`
- `call_pipelined`
- `call_with_timeout`
- `call_without_timeout`

Furthermore, `Redis::SubscribedClient`, `Redis::Distributed`, and
`Redis::Cluster` all wind up using instances of `Redis::Client`
underneath, so all those bases are covered with one relatively
straightforward patch.

I don't think the 2.x line of redis-rb is necessary to add. For one,
it's from 2011, and I think people have had plenty of time to upgrade.
The rubygems.org data bears this out [1]. 2.x also still supports Ruby
1.8 [2], which is well below the beeline's minimum supported Ruby
version.

The hard part about this integration is just the need for extensive
testing. The Redis API is huge, all manner of versions in the 3.x and
4.x lines are still in use, and the redis-rb maintainers have the habit
of adding/modifying commands across patch versions.

For the sake of completeness/paranoia, the tests in this commit have
been run successfully on *every* version of redis-rb between 3.0.0 and
4.1.3 on Ruby 2.6.3 (so that's: 3.0.0, 3.0.1, 3.0.2, 3.0.3, 3.0.4,
3.0.5, 3.0.6, 3.0.7, 3.1.0, 3.2.0, 3.2.1, 3.2.2, 3.3.0, 3.3.1, 3.3.2,
3.3.3, 3.3.5 [note that 3.3.4 was yanked], 4.0.0, 4.0.1, 4.0.2, 4.0.3,
4.1.0, 4.1.1, 4.1.2, and 4.1.3). Because the list is so extensive,
though, I didn't think it was wholly sane to iterate through the
explosion of combinations possible for CircleCI (every Redis version on
every Ruby version...). So I kept the Appraisals simply at the latest
redis 3.x + 4.x gems.

References:

[1] https://ui.honeycomb.io/ruby-together/datasets/rubygems.org/result/J4S6eMgLbv3

[2] https://github.com/redis/redis-rb/tree/be3d7e0b4194fcb7e91bee9153daf6073dbc1268#testing

### CODE REVIEW

**Comment by martin308** [comment]
@findchris that's the plan. I'm working on fixing #41 first, then I'll get to reviewing this

File: lib/honeycomb/integrations/redis.rb [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        elsif char =~ /[[:print:]&&[:ascii:]]/
+          char
+        else
+          bytes = char.unpack("H*").pack("H*").bytes
+          bytes.map { |b| Kernel.format("\\x%02x", b) }.join
+        end
+      end
+
+      def needs_quotes?(char)
+        char =~ /[^[:print:]&&[:ascii:]]|[\\"\n\r\t\a\b' ]/
```
Comment (martin308) [inline lib/honeycomb/integrations/redis.rb]: can we get a comment on this regex?

File: lib/honeycomb/integrations/redis.rb [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+      # @see https://github.com/antirez/redis/blob/0f026af185e918a9773148f6ceaa1b084662be88/src/sds.c#L940-L1067
+      # @see https://github.com/antirez/redis/blob/0f026af185e918a9773148f6ceaa1b084662be88/src/sds.c#L878-L907
+      def escape(char)
+        if ESCAPES.key?(char)
+          ESCAPES[char]
+        elsif char =~ /[[:print:]&&[:ascii:]]/
+          char
+        else
+          bytes = char.unpack("H*").pack("H*").bytes
+          bytes.map { |b| Kernel.format("\\x%02x", b) }.join
```
Comment (martin308) [inline lib/honeycomb/integrations/redis.rb]: can we get a comment on what this is doing?

File: lib/honeycomb/integrations/redis.rb [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+      }.freeze
+
+      # This aims to replicate the algorithm used by redis-cli.
+      #
+      # @see https://github.com/antirez/redis/blob/0f026af185e918a9773148f6ceaa1b084662be88/src/sds.c#L940-L1067
+      # @see https://github.com/antirez/redis/blob/0f026af185e918a9773148f6ceaa1b084662be88/src/sds.c#L878-L907
+      def escape(char)
+        if ESCAPES.key?(char)
+          ESCAPES[char]
+        elsif char =~ /[[:print:]&&[:ascii:]]/
```
Comment (martin308) [inline lib/honeycomb/integrations/redis.rb]: can we get a comment on this regex?

File: spec/honeycomb/integrations/redis_spec.rb [hunk 4]
```diff
@@ -0,0 +1,2192 @@
...
+
+          it "sends just one event" do
+            expect(libhoney_client.events.size).to eq 1
+          end
+
+          it "sends the expected fields" do
+            expect(event_data.first).to include(
+              "meta.span_type" => "root",
+              "redis.command" => "FLUSHDB",
+              "redis.lo
```
[the remark this answers is not in the document]
Comment (martin308) [reply spec/honeycomb/integrations/redis_spec.rb]: I'm not sure if that suggestion will commit, looks like it's having trouble with line numbers but hopefully that's enough to reproduce it at least

[the remark this answers is not in the document]
Comment (ajvondrak) [reply spec/honeycomb/integrations/redis_spec.rb]: Good find! 👍

**Comment by ajvondrak** [comment]
@martin308 Thanks for the review! I maybe went a tad overboard the refactoring/commentary, but it should at least be easier to understand now. Let me know what you think. 😄 

Also gave me a chance to realize that the `unpack` + `pack` combo from before was quite unnecessary, since I was just calling `String#bytes` anyway. 😅

**Pull request merged**